cpu/esp32/periph/spi.c

```diff
@@ -0,0 +1,432 @@
+     * see ESP32 Technical Reference, Section 7.4.1, Table 27
+     * https://www.espressif.com/sites/default/files/documentation/esp32_technical_reference_manual_en.pdf
+     */
+
+    /* hardware CS not used (TODO) */
+    spi_ll_master_select_cs(_spi[bus].periph->hw, INT_MAX);
+    spi_ll_master_set_cs_setup(_spi[bus].periph->hw, 2);
+    spi_ll_master_set_mode(_spi[bus].periph->hw, mode);
+    spi_ll_set_half_duplex(_spi[bus].periph->hw, false);
+
+    int delay_mode = (mode == SPI_MODE_0 || mode == SPI_MODE_3) ? 2 : 1;
+    spi_ll_set_miso_delay(_spi[bus].periph->hw, delay_mode, 0);
+    spi_ll_set_mosi_delay(_spi[bus].periph->hw, 0, 0);
+
+    /*
+     * set SPI clock
+     * see ESP32 Technical Reference, Section 7.8 SPI_CLOCK_REG
+     * https://www.espressif.com/sites/default/files/documentation/esp32_technical_reference_manual_en.pdf
+     */
+
+    /* check whether timing has to be recalculated (time consuming) */
+    if (_spi_clk_freq_hz[clk] != _spi[bus].clk_last) {
+        uint32_t apb_clk = rtc_clk_apb_freq_get();
+        uint32_t clk_reg;
+
+        if (apb_clk / 5 < _spi_clk_freq_hz[clk]) {
+            LOG_TAG_ERROR("spi", "APB clock rate (%"PRIu32" Hz) has to be at "
+                          "least 5 times SPI clock rate (%"PRIu32" Hz)\n",
+                          apb_clk, _spi_clk_freq_hz[clk]);
+            assert(false);
+        }
+
+        /* duty cycle is measured in is 1/256th, 50% = 128 */
+        int _clk = spi_ll_master_cal_clock(apb_clk, _spi_clk_freq_hz[clk],
+                                           128, &clk_reg);
+
+        _spi[bus].clk_last = _spi_clk_freq_hz[clk];
+        _spi[bus].timing.clock_reg = clk_reg;
+        _spi[bus].timing.timing_miso_delay = 0;
+        _spi[bus].timing.timing_dummy = 0;
+
+        DEBUG("%s bus %d: SPI clock frequency: clk=%"PRIu32" eff=%d "
+              "reg=%08"PRIx32"\n",
+              __func__, bus, _spi_clk_freq_hz[clk], _clk, clk_reg);
+    }
+    spi_ll_master_set_clock_by_reg(_spi[bus].periph->hw,
+                                   &_spi[bus].timing.clock_reg);
+
+#if defined(MCU_ESP32C3)
+    /*
+     * If the SPI mode has been changed, the clock signal is only set to the
+     * correct level at the beginning of the transfer on the ESP32C3. However,
+     * if a generic GPIO is used as CS signal instead of the hardware CS,
+     * the CS signal is already LOW at this time. Thus, the clock signal will
+     * have the wrong level when the SPI mode is changed and the CS signal
+     * becomes LOW.
+     * The following is a workaround by receiving a dummy byte without pulling
+     * the CS signal LOW when the mode has been changed.
+     */
+    if (_spi[bus].mode_last != mode) {
+        uint8_t temp = 0xff;
+        spi_transfer_bytes(bus, GPIO_UNDEF, false, &temp, &temp, 1);
+        _spi[bus].mode_last = mode;
+    }
+#elif defined(MCU_ESP32)
+    /* This workaround isn't needed on ESP32 */
+#else
+#error Platform implementation is missing
```

Comment:
This is something we don't know for furture ESP32x SoCs. So I left this error message here as a reminder so that we to have to check this explicitly when RIOT is ported to another ESP32x SoC. With ESP32S2 support it will look like this:
```.c
#if defined(CPU_FAM_ESP32C3) || defined(CPU_FAM_ESP32S3)
    /*
     * If the SPI mode has been changed, the clock signal is only set to the
     * correct level at the beginning of the transfer on the ESP32x3. However,
     * if a generic GPIO is used as CS signal instead of the hardware CS,
     * the CS signal is already LOW at this time. Thus, the clock signal will
     * have the wrong level when the SPI mode is changed and the CS signal
     * becomes LOW.
     * The following is a workaround by receiving a dummy byte without pulling
     * the CS signal LOW when the mode has been changed.
     */
    if (_spi[bus].mode_last != mode) {
        uint8_t temp = 0xff;
        spi_transfer_bytes(bus, GPIO_UNDEF, false, &temp, &temp, 1);
        _spi[bus].mode_last = mode;
    }
#elif defined(CPU_FAM_ESP32) || defined(CPU_FAM_ESP32S2)
    /* This workaround isn't needed on ESP32 and ESP32-S2 */
#else
#error Platform implementation is missing
#endif
```